<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"/><title>remixt/</title><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="alternate" type="application/rss+xml" title="Shah Lab Research" href="http://shahcompbio.github.io/feed.xml"><link href="/css/style.css" rel="stylesheet" media="screen"><link rel="shortcut icon" href="/images/favicon.png"/> <script>!function(e){var t,a={kitId:"xeu8jut",scriptTimeout:3e3},c=e.documentElement,i=setTimeout(function(){c.className=c.className.replace(/\bwf-loading\b/g,"")+" wf-inactive"},a.scriptTimeout),n=e.createElement("script"),o=!1,s=e.getElementsByTagName("script")[0];c.className+=" wf-loading",n.src="//use.typekit.net/"+a.kitId+".js",n.async=!0,n.onload=n.onreadystatechange=function(){if(t=this.readyState,!(o||t&&"complete"!=t&&"loaded"!=t)){o=!0,clearTimeout(i);try{Typekit.load(a)}catch(e){}}},s.parentNode.insertBefore(n,s)}(document)</script> <script>!function(e,a,n,t,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(n),s=a.getElementsByTagName(n)[0],o.async=1,o.src=t,s.parentNode.insertBefore(o,s)}(window,document,"script","//www.google-analytics.com/analytics.js","ga"),ga("create","UA-26244371-2","bedford.io"),ga("send","pageview")</script> <script src="/js/jquery.min.js"></script> <script src="/js/transition.js"></script><script src="/js/collapse.js"></script></head><body><div id="header"> <nav class="navbar navbar-default navbar-static-top" role="navigation"><div class="container"><div class="navbar-header"> <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse-element" aria-expanded="false"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button> <object class="logo navbar-brand" data="/images/logo.svg" type="image/svg+xml"></object> <a class="navbar-brand" href="/">shah lab</a></div><div class="collapse navbar-collapse" id="navbar-collapse-element"><ul class="nav navbar-nav navbar-right"><li> <a href="/research/">research</a></li><li> <a href="/blog/">blog</a></li><li> <a href="/papers/">papers</a></li><li> <a href="/projects/">projects</a></li><li> <a href="/team/">team</a></li></ul></div></div> </nav></div><div class="container"><div class="row"><div class="col-md-12"><div class="bigtitle titlebox"><ol class="breadcrumb"><li>remixt</li></ol></div><p><div class="head"> Clone-specific genomic structure estimation in cancer</div></div></div><div class="bigspacer"></div><div class="row"><div class="col-md-3"><div class="bigspacer"></div><div class="smallhead"> Source code</div><div class="pad-left note"><div class="smallspacer"></div> <i class="fa fa-cog fa-fw"></i> <a class="off" href="https://github.com/amcpherson/remixt">github.com/amcpherson/remixt</a></div><div class="bigspacer"></div><div class="smallhead"> Contributors</div><div class="pad-left note"><div class="smallspacer"></div><div> <a class="off" href="https://github.com/amcpherson"> <img class="pull-left avatar" src="https://avatars0.githubusercontent.com/u/1170234?v=3&s=50"><div class="handlebox" style="padding-left:5px""> amcpherson</div> </a></div></div><div class="bigspacer"></div><div class="smallhead"> Latest commits</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/amcpherson/remixt/commit/2867efa73c7217ad9262381cb6ed8488e608a1ea"> 26 Jun 2017 - <span class="text-gray">meta: added license to conda, bumped version</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/amcpherson/remixt/commit/8b34d5a60ea95b9cb4f4535ed51b47396de09f8a"> 26 Jun 2017 - <span class="text-gray">meta: MIT License</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/amcpherson/remixt/commit/8939ba87f94e6f4717e9046cc5f353ba0b5e3fb9"> 01 Jun 2017 - <span class="text-gray">build: conda version bump</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/amcpherson/remixt/commit/6310e6aaf88e2e49b423e149ffac00617b8e71c0"> 19 May 2017 - <span class="text-gray">minor: optionally dont update h</span> </a></li><div class="smallspacer"></div><li> <i class="fa fa-check-square-o fa-fw"></i> <a class="off" href="https://github.com/amcpherson/remixt/commit/6402986372a239f118e72d6476c4205dbadb9097"> 19 May 2017 - <span class="text-gray">minor: plot label</span> </a></li></ul></div><div class="spacer"></div><div class="smallhead"> Pages</div><div class="pad-left smallnote"><ul class="list-unstyled"><div class="smallspacer"></div><li> <a class="off" href="/projects/remixt/TODO.html"> <i class="fa fa-folder-open-o"></i> / TODO </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/remixt/"> <i class="fa fa-folder-open-o"></i> / </a></li><div class="smallspacer"></div><li> <a class="off" href="/projects/remixt/remixt/states.html"> <i class="fa fa-folder-open-o"></i> / <i class="fa fa-folder-open-o"></i> / states </a></li></ul></div><div class="spacer"></div></div><div class="col-md-8"><div class="post"><p><h1>ReMixT</h1><p>ReMixT is a tool for joint inference of clone specific segment and breakpoint copy number in whole genome sequencing data. The input for the tool is a set of segments, a set of breakpoints predicted from the sequencing data, and normal and tumour bam files. Where multiple tumour samples are available, they can be analyzed jointly for additional benefit.</p><h2>Installation</h2><h3>Installing from conda</h3><p>The recommended method of installation for ReMixT is using <code>conda</code>. First install <a href="https://store.continuum.io/cshop/anaconda/">anaconda python</a> from the continuum website. Then add my channel, and the bioconda channel, and install ReMixT as follows.</p><div class="highlight"><pre><code class="language-" data-lang="">conda config --add channels https://conda.anaconda.org/dranew
conda config --add channels 'bioconda'
conda install remixt
</code></pre></div><h3>Installing from source</h3><h4>Clone Source Code</h4><p>To install the code, first clone from bitbucket. A recursive clone is preferred to pull in all submodules.</p><div class="highlight"><pre><code class="language-" data-lang="">git clone --recursive git@bitbucket.org:dranew/remixt.git
</code></pre></div><h4>Dependencies</h4><p>To install from source you will need several dependencies. A list of dependencies can be found in the <code>conda</code> <code>yaml</code> file in the repo at <code>conda/remixt/meta.yaml</code>.</p><h4>Build executables and install</h4><p>To build executables and install the ReMixT code as a python package run the following command in the ReMixT repo:</p><div class="highlight"><pre><code class="language-" data-lang="">python setup.py install
</code></pre></div><h2>Setup ReMixT</h2><h3>Reference genome</h3><p>Download and setup of the reference genome is automated. The default is hg19. Select a directory on your system that will contain the reference data, herein referred to as <code>$ref_data_dir</code>. The <code>$ref_data_dir</code> directory will be used in many of the subsequent scripts when running destruct.</p><p>Download the reference data and build the required indexes:</p><div class="highlight"><pre><code class="language-" data-lang="">remixt create_ref_data $ref_data_dir
</code></pre></div><h3>Mappability file</h3><p>Additionally, ReMixT requires a mappability file to be generated. We have provided a workflow for generating a mappability file based on <code>bwa</code> alignments, for other aligners, you may want to create your own mappability workflow, see <code>remixt/mappability/bwa/workflow.py</code> as an example.</p><p>To create a mappability file for <code>bwa</code>, run:</p><div class="highlight"><pre><code class="language-" data-lang="">remixt mappability_bwa $ref_data_dir
</code></pre></div><p>Note that this workflow will take a considerable amount of time and it is recommended you run this part of ReMixT setup on a cluster or multicore machine.</p><p>For parallelism options see the section <a href="#markdown-header-parallelism-using-pypeliner">Parallelism using pypeliner</a>.</p><h2>Running ReMixT</h2><h3>Input Data</h3><p>ReMixT takes multiple bam files as input. Bam files should be multiple samples from the same patient, with one bam sequenced from a normal sample from that patient.</p><p>Additionally, ReMixT takes a list of predicted breakpoints detected by paired end sequencing as an additional input.</p><h4>Breakpoint Prediction Input Format</h4><p>The predicted breakpoints should be provided in a tab separated file with the following columns:</p><ul><li><code>prediction_id</code></li><li><code>chromosome_1</code></li><li><code>strand_1</code></li><li><code>position_1</code></li><li><code>chromosome_2</code></li><li><code>strand_2</code></li><li><code>position_2</code></li></ul><p>The first line should be the column names, which should be identical to the above list. Each subsequent line is a breakpoint prediction. The <code>prediction_id</code> should be unique to each breakpoint prediction. The <code>chromosome_</code>, <code>strand_</code> and <code>position_</code> columns give the position and orientation of each end of the breakpoint. The values for <code>strand_</code> should be either <code>+</code> or <code>-</code>. A value of <code>+</code> means that sequence to the right of <code>chromosome_</code>, <code>position_</code> is preserved in the tumour chromosome containing the breakpoint. Conversely, a value of <code>-</code> means that sequence to the left of <code>chromosome_</code>, <code>position_</code> is preserved in the tumour chromosome containing the breakpoint.</p><p>The following table may assist in understanding the strand of a break-end. Note that an inversion event produces two breakpoints, the strand configurations for both are shown. Additionally, for inter-chromosomal events, any strand configuration is possible.</p><table><thead><tr><th>Structural Variation</th><th>Strand of Leftmost Break-End</th><th>Strand of Rightmost Break-End</th></tr></thead><tbody><tr><td>Deletion</td><td>+</td><td>-</td></tr><tr><td>Duplication</td><td>-</td><td>+</td></tr><tr><td>Inversion (Breakpoint A)</td><td>+</td><td>+</td></tr><tr><td>Inversion (Breakpoint B)</td><td>-</td><td>-</td></tr></tbody></table><h3>ReMixT Command Line</h3><p>Running ReMixT involves invoking a single command, <code>remixt run</code>. The result of ReMixT is an <a href="https://www.hdfgroup.org">hdf5</a> file storing <a href="http://pandas.pydata.org">pandas</a> tables.</p><p>Suppose we have the following list of inputs:</p><ul><li>Normal sample with ID <code>123N</code> and bam filename <code>$normal_bam</code></li><li>Tumour sample with ID <code>123A</code> and bam filename <code>$tumour_a_bam</code></li><li>Tumour sample with ID <code>123B</code> and bam filename <code>$tumour_b_bam</code></li><li>Breakpoint table in TSV format with filename <code>$breakpoints</code></li></ul><p>Additionally, ReMixT will generate the following outputs:</p><ul><li>Results as HDF5 file storing pandas tables with filename <code>$results_h5</code></li><li>Temporary files and logs stored in directory <code>$remixt_tmp_dir</code> (directory created if it doesnt exist)</li></ul><p>Given the above inputs and outputs run ReMixT as follows:</p><div class="highlight"><pre><code class="language-" data-lang="">remixt run $ref_data_dir $raw_data_dir $breakpoints \
    --normal_sample_id 123N \
    --normal_bam_file $normal_bam \
    --tumour_sample_ids 123A 123B \
    --tumour_bam_files $tumour_a_bam $tumour_b_bam \
    --results_files $results_h5
    --tmpdir $remixt_tmp_dir
</code></pre></div><p>Note that ReMixT creates multiple jobs and many parts of ReMixT are massively parallelizable, thus it is recommended you run ReMixT on a cluster or multicore machine. For parallelism options see the section <a href="#markdown-header-parallelism-using-pypeliner">Parallelism using pypeliner</a>.</p><h3>Output File Formats</h3><p>The main output file is an HDF5 store containing pandas dataframes. These can be extracted in python or viewed using the ReMixT viewer. Important tables include:</p><ul><li><code>stats</code>: statistics for each restart</li><li><code>solutions/solution_{idx}/cn</code>: segment copy number table for solution <code>idx</code></li><li><code>solutions/solution_{idx}/brk_cn</code>: breakpoint copy number table for solution <code>idx</code></li><li><code>solutions/solution_{idx}/h</code>: haploid depths for solution <code>idx</code></li></ul><h4>Statistics</h4><p>ReMixT uses optimal restarts and model selection by BIC. The statistics table contains one row per restart, sorted by BIC. The table contains the following columns:</p><ul><li><code>idx</code>: the solution index, used to refer to <code>solutions/solution_{idx}/*</code> tables.</li><li><code>bic</code>: the bic of this solution</li><li><code>log_posterior</code>: log posterior of the HMM</li><li><code>log_posterior_graph</code>: log posterior of the genome graph model</li><li><code>num_clones</code>: number of clones including normal</li><li><code>num_segments</code>: number of segments</li><li><code>h_converged</code>: whether haploid depths estimation converged</li><li><code>h_em_iter</code>: number of iterations for convergence of h</li><li><code>graph_opt_iter</code>: number of iterations for convergence of genome graph copy number</li><li><code>decreased_log_posterior</code>: whether the genome graph optimization stopped due to a move that decreased the log posterior</li></ul><h4>Segment Copy Number</h4><p>The segment copy number table adds additional columns to the segment counts table described above, including but not limited to:</p><ul><li><code>major_1</code></li><li><code>minor_1</code></li><li><code>major_2</code></li><li><code>minor_2</code></li></ul><p>The columns refer to the major and minor copy number in tumour clone 1 and 2.</p><h4>Breakpoints Copy Number</h4><p>The breakpoint copy number table contains the following columns:</p><ul><li><code>prediction_id</code></li><li><code>cn_1</code></li><li><code>cn_2</code></li></ul><p>The <code>prediction_id</code> column matches the column of the same name in the input breakpoints file, and specifies for which breakpoint prediction the copy number is being provided. The <code>cn_1</code> and <code>cn_2</code> columns provide the clone specific copy number for tumour clone 1 and 2 respectively.</p><h4>Haploid Depths</h4><p>The haploid depths is a vector of <code>M</code> depths for each of the <code>M</code> clones including the normal. To recover cell mixture proportions, simply normalize <code>h</code>.</p><h3>Extracting Tables as TSV files</h3><p>If preferred, it is possible to extract copy number and metadata in TSV and YaML format. For results file <code>$results_h5</code>, extract segment copy number, breakpoint copy number and meta data to files <code>$cn_table</code>, <code>$brk_cn_table</code>, <code>$meta_data</code> respectively as follows:</p><div class="highlight"><pre><code class="language-" data-lang="">remixt write_results \
    $results_h5 $cn_table $brk_cn_table $meta_data
</code></pre></div><h3>ReMixT Viewer</h3><p>There is an experimental viewer for ReMixT at <code>tools/remixt_viewer_app.py</code>. Bokeh '&gt;0.10.0' is required. To use the viewer app, organize your patient sample results files as <code>./patient_*/sample_*.h5</code>. From the directory containing patient subdirectories, run the bokeh server:</p><div class="highlight"><pre><code class="language-" data-lang="">bokeh-server --script $REMIXT_DIR/tools/remixt_viewer_app.py
</code></pre></div><p>Then navigate to <code>http://127.0.0.1:5006/remixt</code>.</p><h2>Test Dataset for ReMixT</h2><p>A test dataset is provided for providing the ability to run a quick analysis of a small dataset to ensure remixt is working correctly.</p><p>We will assume that the <code>REMIXT_DIR</code> environment variable points to a clone of the ReMixT source code. Additionally, create a directory, and set the environment variable <code>WORK_DIR</code> to the location of that directory.</p><p>First use the <code>remixt create_ref_data</code> sub-command to create a reference dataset. Specify a config, and use the example config that restricts to chromosome 15.</p><div class="highlight"><pre><code class="language-" data-lang="">remixt create_ref_data $WORK_DIR/ref_data \
    --config $REMIXT_DIR/examples/chromosome_15_config.yaml
</code></pre></div><p>Use <code>wget</code> to retrieve a precomputed mappability file.</p><div class="highlight"><pre><code class="language-" data-lang="">wget ftp://ftp.bcgsc.ca/public/shahlab/ReMixT/hg19.100.bwa.mappability.h5 --directory-prefix $WORK_DIR/ref_data/
</code></pre></div><p>Use <code>wget</code> to retrieve the example bam files and their indices for chromosome 15, and the breakpoints file with chromosome 15 breakpoints.</p><div class="highlight"><pre><code class="language-" data-lang="">wget ftp://ftp.bcgsc.ca/public/shahlab/ReMixT/HCC1395*chr15.bam* --directory-prefix $WORK_DIR/
wget ftp://ftp.bcgsc.ca/public/shahlab/ReMixT/HCC1395_breakpoints.tsv --directory-prefix $WORK_DIR/
</code></pre></div><p>Use the <code>remixt run</code> sub-command to run a remixt analysis.</p><div class="highlight"><pre><code class="language-" data-lang="">remixt run $WORK_DIR/ref_data $WORK_DIR/raw_data $WORK_DIR/HCC1395_breakpoints.tsv \
    --config $REMIXT_DIR/examples/chromosome_15_config.yaml \
    --tmpdir $WORK_DIR/tmp_remixt \
    --tumour_sample_ids HCC1395 \
    --tumour_bam_files $WORK_DIR/HCC1395_chr15.bam \
    --normal_sample_id HCC1395BL \
    --normal_bam_file $WORK_DIR/HCC1395BL_chr15.bam \
    --loglevel DEBUG \
    --submit local \
    --results_files $WORK_DIR/HCC1395.h5
</code></pre></div><p>Use the <code>remixt write_results</code> sub-command to write out tables of results and a yaml file containing inferred parameters and other meta data.</p><div class="highlight"><pre><code class="language-" data-lang="">remixt write_results $WORK_DIR/HCC1395.h5 \
    $WORK_DIR/HCC1395_cn.tsv \
    $WORK_DIR/HCC1395_brk_cn.tsv \
    $WORK_DIR/HCC1395_info.yaml
</code></pre></div><p>Finally, create a visualization of the solutions using the <code>remixt visualize_solutions</code> sub-command.</p><div class="highlight"><pre><code class="language-" data-lang="">remixt visualize_solutions $WORK_DIR/HCC1395.h5 \
    $WORK_DIR/HCC1395.html
</code></pre></div><h2>Parallelism Using Pypeliner</h2><p>ReMixT uses the pypeliner python library for parallelism. Several of the scripts described above will complete more quickly on a multi-core machine or on a cluster.</p><p>To run a script in multicore mode, using a maximum of 4 cpus, add the following command line option:</p><div class="highlight"><pre><code class="language-" data-lang="">--maxjobs 4
</code></pre></div><p>To run a script on a cluster with qsub/qstat, add the following command line option:</p><div class="highlight"><pre><code class="language-" data-lang="">--submit asyncqsub
</code></pre></div><p>Often a call to qsub requires specific command line parameters to request the correct queue, and importantly to request the correct amount of memory. To allow correct calls to qsub, use the <code>--nativespec</code> command line option, and use the placeholder <code>{mem}</code> which will be replaced by the amount of memory (in gigabytes) required for each job launched with qsub. For example, to use qsub, and request queue <code>all.q</code> and set the <code>mem_free</code> to the required memory, add the following command line options:</p><div class="highlight"><pre><code class="language-" data-lang="">--submit asyncqsub --nativespec "-q all.q -l mem_free={mem}G"
</code></pre></div><h1>License</h1><p>ReMixT is released under the <a href="http://www.opensource.org/licenses/MIT">MIT License</a>.</p></div></div><div class="col-md-1"></div></div></div><div id="footer"><span style="display:none">foo</span></div></body></html>